doc/dev/kclient.rst

```diff
@@ -0,0 +1,345 @@
+related best-practices.
+
+Step One: build the kernel
+==========================
+
+Configure the kernel:
+
+.. bash::
+
+    make defconfig
+
+
+We now have a kernel config with reasonable defaults for the architecture you're
+building on. The next thing to do is to enable configs which will build Ceph and/or
+provide functionality we need to do testing.
+
+.. bash::
+
+    cat > ~/.ceph.config <<EOF
+    CONFIG_XFS_FS=y
+    CONFIG_CEPH_FS=y
+    CONFIG_CEPH_FSCACHE=y
+    CONFIG_CEPH_FS_POSIX_ACL=y
+    CONFIG_CEPH_FS_SECURITY_LABEL=y
+    CONFIG_CEPH_LIB_PRETTYDEBUG=y
+    CONFIG_DYNAMIC_DEBUG=y
+    CONFIG_DYNAMIC_DEBUG_CORE=y
+    CONFIG_KGDB=y
+    CONFIG_FRAME_POINTER=y
+    CONFIG_KGDB_SERIAL_CONSOLE=y
+    EOF
+
+Note, beyond enabling Ceph-related configs, we are also enabling some useful
+debug configs and XFS (as an alternative to ext4 if needed for our root file
+system).
+
+Now, merge the configs.
+
+
+.. bash::
+
+
+    scripts/kconfig/merge_config.sh .config ~/.ceph.config
+
+
+Finally, build the kernel:
+
+.. bash::
+
+    make -j
+
+
+.. note:: This document does not discuss how to get relevant utilities for your
+          distribution to actually build the kernel, like gcc. Please use your search
+          engine of choice to learn how to do that.
+
+
+Step Two: create a VM
+=====================
+
+A virtual machine is a good choice for testing the kernel client for a few reasons:
+
+* You can more easily monitor and configure networking for the VM.
+* You can very rapidly test a change to the kernel (build -> mount in less than 10 seconds).
+* A fault in the kernel won't crash your machine.
+* You have a suite of tools available for analysis on the running kernel.
+
+The main decision for you to make is what Linux distribution you want to use.
+This document uses Arch Linux due to the author's familiarity. We also use LVM
+to create a volume. You may use partitions or whatever mechanism you like to
+create a block device. In general, this block device will be used repeatedly in
+testing. You may want to use snapshots to avoid a VM somehow corrupting your
+root disk and forcing you to start over.
+
+
+.. bash::
+
+    # create a volume
+    VOLUME_GROUP=foo
+    sudo lvcreate -L 256G "$VOLUME_GROUP" -n $(whoami)-vm-0
+    DEV="/dev/${VOLUME_GROUP}/$(whoami)-vm-0"
+    sudo mkfs.xfs "$DEV"
+    sudo mount "$DEV" /mnt
+    sudo pacstrap /mnt base base-devel vim less jq
+    sudo umount /mnt
+
+Once that's done, we should be able to run a VM:
+
+
+.. bash::
+
+    qemu-system-x86_64 -enable-kvm -kernel $(pwd)/arch/x86/boot/bzImage -drive file="$DEV",if=virtio,format=raw -append 'root=/dev/vda rw'
+
+You should see output like:
+
+::
+
+    VNC server running on ::1:5900
+
+You could view that console using:
+
+
+.. bash::
+
+    vncviewer 127.0.0.1:5900
+
+Congratulations, you have a VM running the kernel that you just built.
+
+
+Step Three: Networking the VM
+=============================
```

Comment:
Locally for me I will use only one VM to do the developing and testing:

1, Build ceph from source.
2, Build and install the kernel from source and then reboot
3, Start ceph cluster by using the 'src/vstart.sh` 
4, Just mount the kclient mountpoints

That's it.

Reply:
Perhaps we should add alternate direction of development on the same machine. However, I'd be cautious about recommending changing your machine's host kernel for testing. Is there something about that you find useful for development (like some kind of debugging)?

Reply:
Ah, I think you're building linux in the VM too, yes?

Reply:
Yeah, I will never do or suggest to install the testing kernel built from source on our own machine's host, which could be risky and will make your node won't start or be stable.

It would better do this in the VMs. For all the above 4 steps will be done in the same single VM.